src/main/java/org/prebid/server/bidder/applogy/ApplogyBidder.java

```diff
@@ -0,0 +1,171 @@
+        final MultiMap headers = resolveHeaders();
+
+        return Result.of(Collections.singletonList(
+                HttpRequest.<BidRequest>builder()
+                        .method(HttpMethod.POST)
+                        .uri(requestUrl)
+                        .headers(headers)
+                        .payload(outgoingRequest)
+                        .body(body)
+                        .build()),
+                errors);
+    }
+
+    private MultiMap resolveHeaders() {
+        return HttpUtil.headers();
+    }
+
+    private ExtImpApplogy parseAndValidateImpExt(Imp imp) {
+        final ExtImpApplogy extImpApplogy;
+        try {
+            extImpApplogy = Json.mapper.convertValue(imp.getExt().get("bidder"), ExtImpApplogy.class);
+        } catch (IllegalArgumentException e) {
+            throw new PreBidException(e.getMessage(), e);
+        }
+
+        if (StringUtils.isBlank(extImpApplogy.getToken())) {
+            throw new PreBidException("token is empty");
+        }
+        return extImpApplogy;
+    }
+
+    private Imp processImp(Imp imp) {
+        Banner banner = imp.getBanner();
+        if (banner != null) {
+            if (banner.getH() == null && banner.getW() == null && CollectionUtils.isNotEmpty(banner.getFormat())) {
+                final Format firstFormat = banner.getFormat().get(0);
+                final Banner modifiedBanner = banner.toBuilder()
+                        .h(firstFormat.getH())
+                        .w(firstFormat.getW())
+                        .build();
+                return imp.toBuilder().banner(modifiedBanner).build();
+            }
+        }
+        return imp;
+    }
+
+    @Override
+    public Result<List<BidderBid>> makeBids(HttpCall<BidRequest> httpCall, BidRequest bidRequest) {
+        if (httpCall.getResponse().getStatusCode() == HttpResponseStatus.NO_CONTENT.code()) {
+            return Result.of(Collections.emptyList(), Collections.emptyList());
+        }
+        try {
+            final BidResponse bidResponse = Json.decodeValue(httpCall.getResponse().getBody(), BidResponse.class);
+            return Result.of(extractBids(httpCall.getRequest().getPayload(), bidResponse), Collections.emptyList());
+
+        } catch (DecodeException e) {
+            return Result.emptyWithError(BidderError.badServerResponse("failed to decode json"));
+        }
+    }
+
+    private static List<BidderBid> extractBids(BidRequest bidRequest, BidResponse bidResponse) {
+        if (bidResponse == null || bidResponse.getSeatbid() == null) {
+            return Collections.emptyList();
+        }
+        return bidsFromResponse(bidRequest, bidResponse);
+    }
+
+    private static List<BidderBid> bidsFromResponse(BidRequest bidRequest, BidResponse bidResponse) {
+        final Map<String, BidType> requestImpIdToBidType = bidRequest.getImp().stream()
+                .collect(Collectors.toMap(Imp::getId, ApplogyBidder::getBidType));
```

Comment:
This is not necessary, just have a method that iterates through `bidrequest.imp` and checks `imp.getId().equals(bid.getIpid())` and then applies the `getBidType` to resolve type.

Also, check the comments to `getBidType` below.